src/backend/compiler/compiler.cc

```diff
@@ -1099,27 +1099,124 @@ extern "C" int seashell_preprocessor_run(struct seashell_preprocessor* preproces
+#endif
+  if(preprocessor->messages.size() <= k)
+    return NULL;
+  return preprocessor->messages[k].file.c_str();
+}
+
+extern "C" int seashell_preprocessor_get_diagnostic_line(struct seashell_preprocessor *preprocessor, int k) {
+  if(preprocessor->messages.size() <= k)
+    return 0;
+  return preprocessor->messages[k].line;
+}
+
+extern "C" int seashell_preprocessor_get_diagnostic_column(struct seashell_preprocessor *preprocessor, int k) {
+  if(preprocessor->messages.size() <= k)
+    return 0;
+  return preprocessor->messages[k].col;
+}
+
+#ifndef __EMSCRIPTEN__
+extern "C" const char *seashell_preprocessor_get_diagnostic_message(struct seashell_preprocessor *preprocessor, int k) {
+#else
+std::string seashell_preprocessor_get_diagnostic_message(struct seashell_preprocessor *preprocessor, int k) {
+#endif
+  if(preprocessor->messages.size() <= k)
+    return NULL;
+  return preprocessor->messages[k].mesg.c_str();
+}
+
+static std::string join(std::vector<std::string> vec, char a) {
+  std::string res = "";
+  for(std::vector<std::string>::iterator it = vec.begin(); it != vec.end(); it++) {
+    res += *it + a;
+  }
+  if(res.length())
+    res.resize(res.length()-1);
+  return res;
+}
+
 /*
   This function contains the logic to resolve a local include using Seashell's
     inclusion rules (ie. question folder, then common, looking at same-named
     .c and .o files)
 */
-static std::string resolve_include(struct seashell_preprocessor *preprocessor, std::string fname) {
-  // TODO do this better sometime in the future
-  fname[fname.length()-1]='c';
-  std::string attempt = preprocessor->project_dir + "/" +
-      preprocessor->question_dir + "/" + fname;
-  //fprintf(stderr, "Trying %s\n", attempt.c_str());
+static std::string resolve_include(struct seashell_preprocessor *preprocessor, std::string fname,
+    std::string currentfile, uint line, uint col) {
+
+  char *orig;
+  orig = strcpy(orig, fname.c_str());
+  char *saveptr;
+  char *seg = strtok_r(orig, "/", &saveptr);
+  std::vector<std::string> path;
+  do {
+    path.push_back(seg);
+  } while(seg = strtok_r(NULL, "/", &saveptr));
+
+  char *ofile;
+  ofile = strdup(path[path.size()-1].c_str());
+  seg = strtok_r(ofile, ".", &saveptr);
+  std::vector<std::string> file;
+  do {
+    file.push_back(seg);
+  } while(seg = strtok_r(NULL, ".", &saveptr));
+
+  // check for include of non .h files
+  if(file[file.size()-1] != "h") {
+    preprocessor->messages.push_back(seashell_diag(false, currentfile, "Included files should have extension .h, instead found '" + fname + "'", line, col));
+  }
+
+  path.insert(path.begin(), preprocessor->question_dir);
+  path.insert(path.begin(), preprocessor->project_dir);
+
+  // attempt question folder .c file
+  file.pop_back();
+  file.push_back("c");
+  path.pop_back();
+  path.push_back(join(file, '.'));
+  std::string attempt = join(path, '/');
   if(std::ifstream(attempt)) return attempt;
```

Comment:
Probably want to use stat(2) here?  http://man7.org/linux/man-pages/man2/stat.2.html